drs_downloader/manager.py

```diff
@@ -0,0 +1,393 @@
+        self.max_simultaneous_object_signers = max_simultaneous_object_signers
+        self.max_simultaneous_downloaders = max_simultaneous_downloaders
+        self.max_simultaneous_part_handlers = max_simultaneous_part_handlers
+        self.disable = not show_progress
+        self.part_size = part_size
+
+    @staticmethod
+    def _parts_generator(size: int, start: int = 0, part_size: int = None) -> Iterator[Tuple[int, int]]:
+        """Determine the start,size for each part
+
+        Args:
+            size: size of file
+            start: offset into file 0 based
+            part_size: desired part size
+
+        Returns:
+            iterator of tuple start, size
+        """
+        while size - start > part_size:
+            yield start, start + part_size
+            start += part_size + 1
+            # start += part_size
+        yield start, size
+
+    async def _run_download_parts(self, drs_object: DrsObject, destination_path: Path) -> DrsObject:
+        """Determine number of parts for signed url and create tasks for each part, run them in batches.
+
+        Args:
+            drs_object: Information about a bucket object
+
+        Returns:
+            list of paths to files for each part, in order.
+        """
+        # create a list of parts
+        parts = []
+        for start, size in self._parts_generator(size=drs_object.size, part_size=self.part_size):
+            parts.append((start, size, ))
+
+        if len(parts) > 1000:
+            logger.error(f'tasks > 1000 {drs_object.name} has over 1000 parts, consider optimization. ({len(parts)})')
+
+        paths = []
+        # TODO - tqdm ugly here?
+        for chunk_parts in \
+                tqdm.tqdm(DrsAsyncManager.chunker(parts, self.max_simultaneous_part_handlers),
+                          total=math.ceil(len(parts) / self.max_simultaneous_part_handlers),
+                          desc="File Download Progress",
+                          leave=False,
+                          disable=self.disable):
+            chunk_tasks = []
+            for start, size in chunk_parts:
+                task = asyncio.create_task(self._drs_client.download_part(drs_object=drs_object, start=start, size=size,
+                                                                          destination_path=destination_path))
+                chunk_tasks.append(task)
+
+            chunk_paths = [
+                await f
+                for f in
+                tqdm.tqdm(asyncio.as_completed(chunk_tasks), total=len(chunk_tasks), leave=False,
+                          desc=f"    * {drs_object.name} Part", disable=self.disable)
+            ]
+            # something bad happened
+            if None in chunk_paths:
+                logger.error(f"{drs_object.name} had missing part.")
+                return drs_object
+
+            paths.extend(chunk_paths)
+
+        drs_object.file_parts = paths
+
+        i = 1
+        filename = f"{drs_object.name}"
```

Comment:
According to the [DRS specification](https://ga4gh.github.io/data-repository-service-schemas/preview/release/drs-1.2.0/docs/#operation/GetObject), the `name` field is not required and may be null.

TNU uses the value of the `name` field if it is not null, and if it is null, it uses the last element of the key path (which is the actual filename) as can be seen here:
https://github.com/DataBiosphere/terra-notebook-utils/blob/b53bb8656d502ecbdbfe9c5edde3fa25bd90bbf8/terra_notebook_utils/drs.py#L248

Reply:
Agreed.  Will address today in a change to this PR, including test.